application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/GetTrialLicenseListener.java

```diff
@@ -65,10 +73,27 @@ public String getName()
     @Override
     public void onEvent(Event event, Object source, Object data)
     {
-        DefaultInstalledExtension extension = (DefaultInstalledExtension) source;
+        List<ExtensionId> extensions = ((JobFinishedEvent) event).getRequest().getProperty("extensions");
+        // Retrieve license updates to be sure that we don't override an existing license.
+        trialLicenseGenerator.updateLicenses();
+
+        for (ExtensionId extensionId : extensions) {
+            if (trialLicenseGenerator.canGenerateTrialLicense(extensionId)) {
+                trialLicenseGenerator.generateTrialLicense(extensionId);
+            }
+
+            // Since a free extension has no license to cover it's dependencies, check to see if there are any paid
+            // dependencies that need a trial license.
+            InstalledExtension installedExtension = installedExtensionRepository.getInstalledExtension(extensionId);
+            Collection<ExtensionDependency> dependencies = installedExtension.getDependencies();
 
-        if (trialLicenseGenerator.canGenerateTrialLicense(extension.getId())) {
-            trialLicenseGenerator.generateTrialLicense(extension.getId());
+            for (ExtensionDependency dependency : dependencies) {
+                ExtensionId dependencyId =
+                    new ExtensionId(dependency.getId(), dependency.getVersionConstraint().getVersion());
+                if (trialLicenseGenerator.canGenerateTrialLicense(dependencyId)) {
+                    trialLicenseGenerator.generateTrialLicense(dependencyId);
+                }
+            }
```

Comment:
You're checking only one level in the dependency tree but we could have licensed extensions many levels below. I think you need to iterate **all** the dependency **tree**, making sure to not iterate the descendants of a node for which you generate a license. Also, you shouldn't treat the root node differently. I would write a generic method that receives an extension id as input (the root) and then call itself recursively for its child nodes if we don't generate a license for the input extension id.